Ops/src/OpsWorkerGenetic.cpp

```diff
@@ -147,66 +143,91 @@ void WorkerGenetic::setMutationProbability(double value)
     mMutationProbability = value;
 }
 
+//! @brief Set elites paremter
+void WorkerGenetic::setNumberOfElites(size_t value)
+{
+    mNumElites = value;
+}
+
 
 //! @brief Auxiliary function for selecting parents for crossover
 void WorkerGenetic::selectParents() {
-    double bestValue = std::numeric_limits<double>::max();
-    for(size_t i=0; i<mNumPoints; ++i) {
-        double value = mObjectives[i]*opsRand();
-        if(value < bestValue) {
-            mParent1 = i;
-            bestValue = value;
+    std::vector<size_t> sortedIds = getIdsSortedFromWorstToBest();
+
+    // Select parent 1
+    std::vector<size_t> competitors;
+    competitors.push_back(opsRand()*(mNumPoints));
+    competitors.push_back(opsRand()*(mNumPoints));
+    while(competitors.size() < 2) {
+        competitors.push_back(opsRand()*(mNumPoints));
+    }
+    size_t bestCompetitor = 0;
+    for(auto competitor : competitors) {
+        if(competitor > bestCompetitor) {
+            bestCompetitor = competitor;
         }
     }
-
-    bestValue = std::numeric_limits<double>::max();
-    for(size_t i=0; i<mNumPoints; ++i) {
-        double value = mObjectives[i]*opsRand();
-        if(value < bestValue && i != mParent1) {
-            mParent2 = i;
-            bestValue = value;
+    int bestId = sortedIds[bestCompetitor];
+    mParent1 = bestId;
+
+    // Select parent 2
+    competitors.clear();
+    competitors.push_back(opsRand()*(mNumPoints));
+    competitors.push_back(opsRand()*(mNumPoints));
+    while(competitors.size() < mObjectives.size()/3.0) {
+        competitors.push_back(opsRand()*(mNumPoints));
+    }
+    bestCompetitor = 0;
+    for(auto competitor : competitors) {
+        if(competitor > bestCompetitor) {
+            bestCompetitor = competitor;
         }
     }
+    bestId = sortedIds[bestCompetitor];
+    mParent2 = bestId;
 }
 
 
 //! @brief Auxiliary function for perform crossover
 void WorkerGenetic::crossOver()
 {
-    for(size_t i=0; i<mNumParameters; ++i)
+    mCandidatePoints.push_back(mPoints[mParent1]);
+    mCandidatePoints.push_back(mPoints[mParent2]);
+    double doCrossOver = opsRand();
+    if(doCrossOver < mCrossoverProbability)
     {
-        double x1 = mPoints[mParent1][i];
-        double x2 = mPoints[mParent2][i];
 
-        double doCrossOver = opsRand();
-        if(doCrossOver > mCrossoverProbability)
-        {
-            mChild1[i] = x1;
-            mChild2[i] = x2;
-        }
-        else
+        for(size_t i=0; i<mNumParameters; ++i)
         {
+            double x1 = mPoints[mParent1][i];
+            double x2 = mPoints[mParent2][i];
+
             double mean = (x1+x2)/2.0;
-            double dev = fabs(x2-x1)/2.0;
+            double dev = fabs(x2-x1)/10.0;
```

Comment:
What is this magic value 10.0 that was previously 2.0

Reply:
It is the standard deviation of the normal distribution used for crossover. It gives a slightly higher chance of producing offspring close to the average value of the parents. The value 10 just makes it a little narrower. Perhaps it should be a tunable parameter, but it would be quite tricky for the user to find a decent value anyway.

Reply:
hmm, yes it should probably be tunable, but I guess its not critical right now